Remove redundant shuffle in KnapsackSolver

madeken opened the pull request:
It's a bit simpler to unconditionally shuffle once in `SelectCoins` instead of in `SelectCoinsMinConf` which is repeatedly called.

Empact commented:
NACK 4ce190a0157837df9a4a804f8e3f2941d1ffacf5

https://github.com/bitcoin/bitcoin/blob/c83442e17412b96f00bc29d30763b28184c95f1e/src/wallet/test/coinselector_tests.cpp#L507-L517

https://github.com/bitcoin/bitcoin/blob/c83442e17412b96f00bc29d30763b28184c95f1e/src/wallet/test/coinselector_tests.cpp#L530-L540

madeken commented:
@Empact the comment is wrong...

The reason it sometimes fails and sometimes succeeds is because of the randomization here:
https://github.com/bitcoin/bitcoin/blob/e21129e78e38935eb3310207c5be7cf2b1f422e7/src/wallet/coinselection.cpp#L195

but note at that stage it's iterating over a _sorted_ list:

https://github.com/bitcoin/bitcoin/blob/e21129e78e38935eb3310207c5be7cf2b1f422e7/src/wallet/coinselection.cpp#L255

MarcoFalke commented:
The tests are failing, so they need to be updated in the same commit

gmaxwell commented:
What happens if there are ties in the sort?

madeken commented:
> What happens if there are ties in the sort or multiple smallest elements >= N?

@gmaxwell Ties in the sort are probably not a big deal, because the `ApproximateBestSubset` has it's own bit of randomization. The two cases that have a change in behavior would be if there's a tie `= N`  or the "the sum of all elements is < N is < N, and there is a a tie in the smallest elements `>= N`". In the case of a shuffle would be fully random, while without would be "use the first".

It definitely seems like an odd thing to worry too much about, considering it's the difference between: "This is the first, smallest utxo we have greater than >N" and "This is a randomized, smallest utxo we have greater than >N".

However if it's concerning, a hack to simulate this could be replace the current `Shuffle` with a `std::sort` and then remove the later `std::sort`. Considering that `std::sort` is not stable, it's probably good enough "randomization" for the tie-case.

madeken commented:
@gmaxwell Actually it's pretty irrelevant because `groups` is  *already* shuffled. So this just stops reshuffling it each time `SelectCoinsMinConf` is called.  So in the case of ties, the order is already randomized:

https://github.com/bitcoin/bitcoin/blob/c83442e17412b96f00bc29d30763b28184c95f1e/src/wallet/wallet.cpp#L2531-L2533

madeken commented:
@gmaxwell I'm doubt it matters, but fair enough point. That's easy enough to fix with:  856b0a7bd015551d166cb8001ac45460e659b28d  to just unconditionally shuffle  before all the calls to `SelectCoinsMinConf`

madeken commented:
Rebased, updated original post (description) and removed the test of randomness in `SelectCoinsMinConf` as it's no longer applicable and `SelectCoins` doesn't make it easy to do the same test.

madeken committed:
Lift the wallet shuffle from SelectCoinsMinConf to SelectCoins

That way it's only done once

Empact commented:
That looks better to me, but if deterministic output is significant to privacy, I'd rather have less efficient operation covered by a test than more efficient operation without one, particularly wrt coin selection.

madeken commented:
@Empact even without the shuffle (which it has), the privacy implications are tiny. It only makes a difference in an extreme edge case:  if BnB fails (does its own sorting), and if the sum of < N  is  <N fails (has its own randomization) and if the smallest >= N input has ties.

And even then, the privacy difference between "This is the smallest >= N input (first tie-breaking)" and "This is a smallest >= N input (random tie-breaking)" is going pretty negligible.

gmaxwell commented:
Why bother with the change however?  I'm not really seeing the motivation for it, saving say 0.5ms in a 500ms operation isn't particularly interesting.  Actually figuring out if the privacy difference is material isn't free....

laanwj commented:
Agree with @gmaxwell here.

I'd like to see significant performance numbers before considering this change. I'd suspect an extra shuffle is almost free, and the human work of reviewing this is considerable and likely better spent on other PRs.
